feat(zones): Migrate zones to APIv3 MAASENG-3405

abuyukyi101198 opened the pull request:
## Done

- Created a mock base url environment variable for query testing
- Updated the `hey-api` codegen files
- Refactored `useWebsocketAwareQuery` function to match the React Query types presented by the `hey-api` codegen
- Migrated existing list and get zones query functions to APIv3, added create, update and delete queries
- Implemented the new zones query functions, completely replacing the APIv2 implementation
- Created new test setup code including MSW
- Refactored tests utilizing the zone API to switch to MSW, and add appropriate waiting for the mock server
- Fixed the false-positive test in `ComposeFormFields.test.tsx` _(refer to Notes)_

## QA steps

- [ ] Verify all the zone use cases work as expected
- [ ] Review the new test setup for [MAASENG-4386](https://warthogs.atlassian.net/browse/MAASENG-4386)

The PR is quite large since a lot of tests were migrated. Here are some files to pay specific attention to:
- `api/query/base.ts`
- `testing/utils.tsx`
- `api/query/zones.ts & zones.test.ts`
- `zones/views/ZonesList/ZoneListTable` all components (as implementation examples)
- any test file (as new test utility example)

## Fixes

Resolves: 

[MAASENG-4261](https://warthogs.atlassian.net/browse/MAASENG-4261)
[MAASENG-4262](https://warthogs.atlassian.net/browse/MAASENG-4262)
[MAASENG-4263](https://warthogs.atlassian.net/browse/MAASENG-4263)
[MAASENG-4384](https://warthogs.atlassian.net/browse/MAASENG-4384)
[MAASENG-4385](https://warthogs.atlassian.net/browse/MAASENG-4385)

## Notes

The `"shows warnings if available cores/memory is less than the default"` test in `ComposeFormFields.test.tsx` was passing before the APIv3 implementation, because the the expected display message was being rendered on mount while the schema validation required waiting to change the display message. Prior to v3, tests do not wait, meaning the initial state of the rendered component may pass the test, although it later gets updated and should fail. This false-positive became apparent after the v3 migration, and was also fixed with this PR.


[MAASENG-4386]: https://warthogs.atlassian.net/browse/MAASENG-4386?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAASENG-4261]: https://warthogs.atlassian.net/browse/MAASENG-4261?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAASENG-4262]: https://warthogs.atlassian.net/browse/MAASENG-4262?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAASENG-4263]: https://warthogs.atlassian.net/browse/MAASENG-4263?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://maas-ui-5596.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/13465)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

ndv99 reviewed (changes requested):
There's only a few things that I think really need changing, but my biggest gripe here is the exporting of the mock resolvers from the zones test file - see below for details.

Other than that, this is really good! I especially like the `renderWithProviders` util, I think that's so much cleaner than what existed before it

abuyukyi101198 commented on src/app/api/query/base.ts:
```diff
@@ -46,65 +44,38 @@ export const useWebSocket = () => {
-        const mappedKey = wsToQueryKeyMapping[model];
-        const modelQueryKey = queryKey[0];
+    return subscribe(() => {
+      const mappedKey = "zones"; // TODO: add support for other endpoints
```
That subscribe method body was that way in the previous implementation. Removing the conditional invalidation does not seem to change anything, so changed the subscription to invalidate the queries without condition 👍

abuyukyi101198 commented on src/testing/utils.tsx:
```diff
@@ -559,3 +514,72 @@ export const renderWithHistoryRouter = (
 };
+
+/* New utils with easier use */
+export const BASE_URL = import.meta.env.VITE_APP_MAAS_URL;
```
We could discuss that with [MAASENG-4386](https://warthogs.atlassian.net/browse/MAASENG-4386)

abuyukyi101198 commented on src/app/zones/views/ZonesList/ZonesList.test.tsx:
```diff
@@ -1,31 +1,40 @@
-    renderWithBrowserRouter(<ZonesList />, { route: "/zones", queryData });
+  it("shows a zones table if there are any zones", async () => {
+    renderWithBrowserRouter(<ZonesList />, { route: "/zones" });
+    await waitFor(() => expect(zoneResolvers.listZones.resolved).toBeTruthy());
```
You are right, not in this test. But in others waiting is necessary, I must have assumed this test also needed waiting. Apparently not the case

abuyukyi101198 commented on src/app/api/query/zones.test.ts:
```diff
@@ -1,70 +1,307 @@
+  typeof data.name === "string" &&
+  (!data.description || typeof data.description === "string");
+
+export const zoneResolvers = {
```
I did move resolvers into their own directory under `src/testing`, however, I prefer to keep the resolvers as props of a `zoneResolvers` object because of the `resolved` props. Some tests need a way of checking if the queries are resolved, and the components they render have no way of telling if the queries are successful. So `resolved` are needed until a better solution is found.

ndv99 commented on src/app/base/components/ZoneSelect/ZoneSelect.test.tsx:
```diff
@@ -14,13 +14,7 @@ import {
-afterEach(() => {
-  mockServer.resetHandlers();
-});
-afterAll(() => mockServer.close());
```
I didn't actually know that these weren't needed - the tests seem to be passing without them!

abuyukyi101198 replied:
They are moved into the setupServer function in src/testing/utils.tsx! I was also worried I couldn't move these outside the test file, but apparently they are simply subscription functions that work anywhere

ndv99 commented on src/testing/resolvers/zones.ts:
```diff
@@ -0,0 +1,147 @@
+  listZones: {
+    resolved: false,
+    handler: () =>
+      http.get(`${BASE_URL}MAAS/a/v3/zones_with_summary`, () => {
```
I feel like we could maybe keep track of API endpoint URLs in a const or something, but this is beyond the scope here. I'll add this as a comment to [MAASENG-4386](https://warthogs.atlassian.net/browse/MAASENG-4386).

ndv99 commented on src/testing/resolvers/zones.ts:
```diff
@@ -0,0 +1,147 @@
+        try {
+          const data = await request.json();
+          if (!isValidZoneRequest(data)) return HttpResponse.error();
+          const newZone = {
+            id: mockZones.items.length + 1,
+            name: data.name,
+            description: data.description ?? "",
+            controllers_count: 0,
+            devices_count: 0,
+            machines_count: 0,
+          };
+          mockZones.items.push(newZone);
+          zoneResolvers.createZone.resolved = true;
```
I feel like this might be overkill, we could probably just return HTTP 201 without actually doing anything, like we do in Site Manager. But I respect the effort, and it's not causing any issues either, so I'm happy for it to stay. Same goes for the handlers on the other methods in this file

abuyukyi101198 replied:
I wanted to have the mock server resemble an actual backend as much as possible 😄

ndv99 reviewed (commented):
Code is looking really good! I did notice something in QA though; if you edit a zone, the list updates immediately with the new edits, as it should. But if you edit it again, it still shows the old info, before your edits. Probably the query key for the single zone fetch is not being invalidated

abuyukyi101198 commented:
Fixed the `EditZone` autofill issue, it seems that since the query invalidation was in the hook, and `closeForm` was unmounting the component the hook was in before invalidation could complete, it would not preperly refetch the zone on a second edit.

I moved the invalidation logic to the form's `onSuccess` instead, which should resolve the issue (also did the same for `DeleteZone`)

ndv99 reviewed (approved):
LGTM - Congrats! Our first view fully migrated to V3, this is a big milestone :tada:
